src/services/pathfinding.service.ts

```diff
@@ -1,54 +1,164 @@
+ * @param start
+ * @param end
  */
 export const findPathOnFloor = (
   floorId: number,
   start: Location,
   end: Location
 ): Line[] => {
-  return [];
+  const { travelNodes } = buildingFloors.find(({ id }) => id === floorId);
+  const nodes = _.cloneDeep(travelNodes);
+
+  const edges: Line[] = traverseNodes(nodes);
+  const startEdge: Line = findClosestLine(edges, start);
+  const endEdge: Line = findClosestLine(edges, end);
+
+  const initial: TravelNode = {
+    id: nodes.length,
+    location: start,
+    children: [startEdge[0].id, startEdge[1].id]
+  };
+  const goal: TravelNode = {
+    id: nodes.length + 1,
+    location: end,
+    children: [endEdge[0].id, endEdge[1].id]
+  };
+
+  if (_.isEqual(startEdge, endEdge)) {
+    initial.children.push(goal.id);
+    goal.children.push(initial.id);
+  }
+  nodes.push(initial, goal);
+
+  nodes[startEdge[0].id].children.push(initial.id);
+  nodes[startEdge[0].id].children = nodes[startEdge[0].id].children.filter(
+    c => startEdge[1].id !== c
+  );
+
+  nodes[startEdge[1].id].children.push(initial.id);
+  nodes[startEdge[1].id].children = nodes[startEdge[1].id].children.filter(
+    c => startEdge[0].id !== c
+  );
+
+  nodes[endEdge[0].id].children.push(goal.id);
+  nodes[endEdge[0].id].children = nodes[endEdge[0].id].children.filter(
+    c => endEdge[1].id !== c
+  );
+
+  nodes[endEdge[1].id].children.push(goal.id);
+  nodes[endEdge[1].id].children = nodes[endEdge[1].id].children.filter(
+    c => endEdge[0].id !== c
+  );
+
+  const searchPath = search(nodes, initial, goal);
+
+  if (searchPath === null) return null;
+
+  const shortestPath: Line[] = [];
+  let { id } = goal;
+  while (id !== initial.id) {
+    const parentId: number = searchPath[id];
+    shortestPath.push([nodes[id], nodes[parentId]]);
+    id = parentId;
+  }
+  return shortestPath;
+};
+
+/**
+ * A* implementation for indoor pathfinding
+ * @param nodes
+ * @param initial
+ * @param goal
+ */
+const search = (
+  nodes: TravelNode[],
+  initial: TravelNode,
+  goal: TravelNode
+): {} => {
+  const closed = {};
+  const open: PriorityQueue<PQItem> = new PriorityQueue<PQItem>({
+    comparator: (a, b) => {
+      return f(nodes[a.id], goal, a.g) - f(nodes[b.id], goal, b.g);
+    }
+  });
+  open.queue({ id: initial.id, parent: null, g: 0 });
+  while (open.length > 0) {
+    const current = open.dequeue();
+    if (!(current.id in closed)) {
+      closed[current.id] = current.parent;
+      if (current.id === goal.id) return closed;
+      nodes[current.id].children.forEach(child => {
+        if (child in closed) return;
+        const g =
+          current.g +
+          getDistance(nodes[current.id].location, nodes[child].location);
+        open.queue({ id: child, parent: current.id, g });
+      });
+    }
+  }
+  return null;
+};
+
+/**
+ * The sum of the distance from the destination to a node and the distance from the start location to a node
+ * @param travelNode
+ * @param goal
+ * @param g
```

Comment:
What is g, can you add a description of the parameter. I initially thought it was `goal` but I see that's already another parameter.

Reply:
![image](https://user-images.githubusercontent.com/36641869/77602361-a8fce900-6ee3-11ea-8d82-0ed6eec4aa26.png)
As described in https://en.wikipedia.org/wiki/A*_search_algorithm, " n is the next node on the path, g(n) is the cost of the path from the start node to n, and h(n) is a heuristic function that estimates the cost of the cheapest path from n to the goal.". Would you like us to write this down as a comment?

Reply:
Yes I think it would be helpful to explain the parameter when the name is not descriptive. You can do it like in java this is an example from the java doc styleguide `@param text  the string to display.  If the text is null, the tool tip is turned off for this component.`